feat: app accent color

likhithpraveenk opened the pull request:
<!-- Just a reminder that text in these brackets is not visible in the rendered output.
You also do **not** need to use this template. You can remove and modify anything however you like!
Its just a suggestion :)
-->

## Changes
<!-- Please describe what this Pull request adds or changes. Possibly with images -->
Adds Accent Color setting which generates colorscheme

## Todo before merging
<!-- Add custom todos if deemed necessary to give others an overview on how far this PR is progressing -->

<!-- Did you add UI text? 
No? Just check or remove the box
Yes? Make sure you replace hardcoded strings with translations -->
- [x] Translations
<!-- Did you add settings to the settings screen?
No? Just check or remove the box 
Yes? Make sure they are resettable using the button on the respective page! (see finamp_settings_helper.dart)-->
- [x] Reset Settings
<!-- Did you run into problems?
No? Just check or remove the box
Yes? Look at the CONTRIBUTING.md, maybe you can add something to help others! -->

## Related Issues
<!-- List relevant issues to this PR using a format like this
- closes #xyz 
- fixes #abc
- follow up to #pr-name
-->
feat #919
This is a small feature and the accent obviously won't override the dynamic now playing screen elements. Is that required? I'm not very good at UI/UX so suggestions are welcome.


<img width="200" height="200" alt="Screenshot_1759958587" src="https://github.com/user-attachments/assets/679ca959-29a3-4aa7-b2c9-92b2a7eaf0ed" />

<img width="200" height="200" alt="Screenshot_1759958598" src="https://github.com/user-attachments/assets/6d70d45b-1370-41c2-b482-5572b4c0fb07" />
Bottom Sheet

likhithpraveenk committed:
add app accent color

flloschy commented:
I would reorganize the layout to something like this
```
 /'''\   |''''''''''''''''''''|
|color|  | #009688            |
 \___/   |____________________|
```

Where the circle is clickable to select preset or saved colors but also shows the current HEX value, and then this could be placed into the settings screen directly without an popup :)

Also a warning if contrast ratio is too bad could be neat, kinda like realtimecolors.com does. though idk how that calculation actually works

likhithpraveenk commented:
I could remove the bottom sheet and place everything in the settings tile but there would be no space to mention "Accent Color" or something of that sort. 

Regarding contrast ratio, flutter generates the color scheme based on given color [Color scheme.fromSeed](https://api.flutter.dev/flutter/material/ColorScheme/ColorScheme.fromSeed.html)

Chaphasilor commented on lib/services/accent_color_helper.dart:
```diff
@@ -0,0 +1,54 @@
+import 'package:hive_ce_flutter/hive_flutter.dart';
+import 'package:hive_ce_flutter/adapters.dart';
+
+/// A helper class for persisting and retrieving the app's accent color
```
is there a reason why this isn't part of `finamp_models.dart`?

Chaphasilor replied:
I guess it's because the theme needs to wrap the entire app. But maybe you can use a different approach than adding a `ValueListenableBuilder`:

1. Just use a the `finampSettingsProvider` and watch the specific accent color setting, i.e. `ref.watch(finampSettingsProvider.accentColor)`, then pass that value to `getColorScheme()`

2. Or an approach similar to this one:  
https://github.com/jmshrv/finamp/blob/493078033c098cb2668f7facfc4a13ac072caa35/lib/main.dart#L346-L348

likhithpraveenk replied:
I'm not sure on how to use the ref.watch without moving the MaterialApp to a ConsumerWidget. I could use a listener and a local variable but that requires set state and obviously its not a good idea to use set state in **_Finamp_** root app. 
Another idea is to migrate the hive box ThemeMode to dynamic and store our accent color in it.

Komodo5197 replied:
You can use the Consumer widget for this.

Komodo5197 replied:
If you feel like cleaning things up you could also add providers in the LocaleHelper and ThemeModeHelper classes and then watch them in the same consumer, eliminating the need for the ValueListenableBuilders.

```
static Provider<Locale?> localeProvider = Provider((ref) {
    final listener = localeListener;
    void callback() => ref.state = locale;
    listener.addListener(callback);
    ref.onDispose(() => listener.removeListener(callback));
    return locale;
  });
```

```
static Provider<ThemeMode> themeModeProvider = Provider((ref) {
    final listener = themeModeListener;
    void callback() => ref.state = listener.value.get("ThemeMode") ?? ThemeMode.system;
    listener.addListener(callback);
    ref.onDispose(() => listener.removeListener(callback));
    return listener.value.get("ThemeMode") ?? ThemeMode.system;
  });
```

Chaphasilor commented on lib/services/accent_color_helper.dart:
```diff
@@ -0,0 +1,54 @@
+    final green = g.toRadixString(16).padLeft(2, '0');
+    final blue = b.toRadixString(16).padLeft(2, '0');
+
+    return '${includeAlpha ? alpha : ''}$red$green$blue'.toUpperCase();
```
You should add this method to our existing extensions for `Color`, where `atContrast()` and such are defined:
https://github.com/jmshrv/finamp/blob/redesign/lib/at_contrast.dart

You could add the `toHex()` extension below and rename the file to be more generic.

Also, it might be possible to simplify your implementation:
```dart
"#${(Colors.amber.toARGB32() & 0xFFFFFF).toRadixString(16).padLeft(6, '0')}"
```
could work (not tested), or check out this answer:
https://stackoverflow.com/a/61186997

Same for the other helper method.

Chaphasilor commented on lib/color_schemes.g.dart:
```diff
@@ -84,3 +84,14 @@ const darkColorScheme = ColorScheme(
+/// If [color] is provided -> returns a generated color scheme
+/// otherwise falls back to default color schemes
+/// [lightColorScheme] or [darkColorScheme]
+ColorScheme getColorScheme(Color? color ,{bool dark = false}) {
```
Please just pass the `Brightness` as a required parameter, instead of `dark` as a boolean.

Chaphasilor commented on lib/components/LayoutSettingsScreen/accent_color_selector.dart:
```diff
@@ -0,0 +1,152 @@
+    final controller = TextEditingController(text: AccentColorHelper.toHex(currentColor));
+    Color? previewColor = currentColor ?? Colors.transparent;
+
+    final suggestedColors = [Colors.pink, Colors.teal, Colors.amber, Colors.cyan];
```
Not sure how useful these suggestions are. Might also be nice to show previously selected colors here (those that were actually confirmed/saved).

Chaphasilor commented on lib/components/LayoutSettingsScreen/accent_color_selector.dart:
```diff
@@ -0,0 +1,152 @@
+                    ),
+                    alignment: Alignment.center,
+                    child: previewColor == null ? Text(AppLocalizations.of(context)!.invalidColorCode) : null,
+                  ),
```
I'm missing a proper color picker here. From the screenshot I assumed this container could be clicked to open an RGB or HSL color picker, but that's not the case. Only being able to choose form presets or manually enter hex codes seems a bit limited.  
I'm not opposed to depending on another package for this.

Alternatively (or additionally, depending on what the package offers), it would be nice to have HSL (hue, saturation, lightness) sliders to control the color more deliberately. I find that RGB "rainbow pickers" are not great for making small adjustments to the selected color.

Chaphasilor reviewed (commented):
Hey, thanks for the PR!

Those changes already look pretty good; I've left a few comments on things that could be improved.  
I like the approach of using a bottom sheet to configure this setting, keeping the settings list tile height at a single line.

Chaphasilor commented:
Oh and one additional thing:  
It would of course be awesome to base the color theme dynamically on the device theme (e.g. Material You on Android, not sure if there's now an equivalent for iOS).  
This is completely **optional** for this PR though, I'm just mentioning it for completeness' sake, in case you're interested.

likhithpraveenk commented:
Thanks for good feedback. I'll make the suggested changes.

likhithpraveenk committed:
refactor: accent color

likhithpraveenk committed:
moved onTap to color circle

likhithpraveenk committed:
refactor: themeMode and locale from hive boxes to settings

likhithpraveenk commented:
@Chaphasilor I think i made all the required changes. Adding HSL picker is the only thing left. [flutter_colorpicker](https://pub.dev/packages/flutter_colorpicker) this looks promising.

<img width="200" height="200" alt="Screenshot_1760023800" src="https://github.com/user-attachments/assets/28aa54a3-287f-4d06-9544-e5b70e3a60f0" />
<img width="200" height="200" alt="Screenshot_1760023817" src="https://github.com/user-attachments/assets/a3e394fb-919a-4ac2-b693-72dc722b2c2b" />


@Komodo5197 Yeah so i cleaned up the themeMode and locale to settings as well. I did the following minimal testing with them and they work
- changed device theme and in settings 
- changed device language and in settings 
- reset settings

Komodo5197 commented on lib/components/LanguageSelectionScreen/language_list.dart:
```diff
@@ -1,20 +1,21 @@
 import '../padded_custom_scrollview.dart';
 
-class LanguageList extends StatefulWidget {
+class LanguageList extends ConsumerStatefulWidget {
```
It looks like locales never changes, so LanguageList can be converted to a stateless widget with locales calculated during build.

Komodo5197 commented on lib/components/LanguageSelectionScreen/language_list.dart:
```diff
@@ -25,61 +26,54 @@ class _LanguageListState extends State<LanguageList> {
-            // const stops it from switching when going to/from the same
-            // language as the system language (e.g., system to English on a
-            // device set to English)
+    ref.watch(finampSettingsProvider.locale);
```
This shouldn't be needed, as the LanguageListTiles now watch the setting on their own.  This also means the const constructor exception shouldn't be needed, as that was presumably there to make sure this valueListener rebuild propagated down to the no argument child.

Komodo5197 commented on lib/screens/settings_screen.dart:
```diff
@@ -189,7 +188,9 @@ class _SettingsScreenState extends ConsumerState<SettingsScreen> {
             title: Text(AppLocalizations.of(context)!.language),
-            subtitle: Text(LocaleHelper.locale?.nativeDisplayLanguage ?? AppLocalizations.of(context)!.system),
+            subtitle: Text(
+              ref.read(finampSettingsProvider.locale)?.nativeDisplayLanguage ?? AppLocalizations.of(context)!.system,
```
I wouldn't expect this to actually change, but this should be ref.watch instead of ref.read.

Komodo5197 commented on lib/utils/locale_helper.dart:
```diff
@@ -0,0 +1,18 @@
+import 'package:get_it/get_it.dart';
+
+// Only to be used when you have no WidgetRef
+Locale? getLocale() {
```
I don't think this function is needed, you should be able to just do FianmpSettingsHelper.finampSettings.locale if you don't want reactivity.

Komodo5197 commented on lib/services/widget_bindings_observer_provider.dart:
```diff
@@ -56,8 +54,9 @@ class _FinampProviderBuilderState extends ConsumerState<FinampProviderBuilder> w
   void didChangePlatformBrightness() {
-    var theme = switch (Hive.box<ThemeMode>("ThemeMode").get("ThemeMode")) {
-      null || ThemeMode.system => View.of(context).platformDispatcher.platformBrightness,
+    final themeMode = ref.read(finampSettingsProvider.themeMode);
```
I'd prefer FinampSettingsHelper.finampsettings.themeMode

Komodo5197 commented on lib/services/widget_bindings_observer_provider.dart:
```diff
@@ -41,7 +40,6 @@ class _FinampProviderBuilderState extends ConsumerState<FinampProviderBuilder> w
   void initState() {
     super.initState();
     WidgetsBinding.instance.addObserver(this);
-    ThemeModeHelper.themeModeListener.addListener(didChangePlatformBrightness);
```
We still want to react to themeMode changes.

`ref.listenManual(finampSettingsProvider.themeMode,(_,_)=>didChangePlatformBrightness());`

Komodo5197 commented on lib/main.dart:
```diff
@@ -269,13 +267,12 @@ Future<void> setupHive() async {
     Hive.openBox<FinampSettings>("FinampSettings", path: dir.path),
-    Hive.openBox<ThemeMode>("ThemeMode", path: dir.path),
     Hive.openBox<FinampStorableQueueInfo>("Queues", path: dir.path),
-    Hive.openBox<Locale?>(LocaleHelper.boxName, path: dir.path),
```
I do like moving the theme and locale settings into the standard settings object, but I think we need to add a migration.

likhithpraveenk replied:
I'm not familiar with migration very much. is it something like this?
```dart
Future<void> _migrateThemeModeLocale() async {
  Box<ThemeMode> oldThemeModeBox = Hive.box<ThemeMode>("ThemeMode");
  Box<Locale?> oldLocaleBox = Hive.box<Locale?>("Locale");

  final finampSettings = FinampSettingsHelper.finampSettings;

  bool needsUpdate = false;

  if (oldThemeModeBox.isNotEmpty) {
    var oldTheme = oldThemeModeBox.get("ThemeMode");
    if (oldTheme != finampSettings.themeMode) {
      finampSettings.themeMode = oldTheme!;
      needsUpdate = true;
    }
  }

  if (oldLocaleBox.isNotEmpty) {
    var oldLocale = oldLocaleBox.get("Locale");
    if (oldLocale != finampSettings.locale) {
      finampSettings.locale = oldLocale;
      needsUpdate = true;
    }
  }

  if (needsUpdate) {
    FinampSettingsHelper.overwriteFinampSettings(finampSettings);
  }

  await oldThemeModeBox.deleteFromDisk();
  await oldLocaleBox.deleteFromDisk();
}
```

Komodo5197 replied:
The way we've been doing most of the migrations is to just have a flag for them.  So you create a new bool setting like hasCompletedThemLocaleMigration which has a DefaultSetting of true, but has false in the HiveField default ( see hasCompletedDownloadsFileOwnerMigration for an example).  And then if that's false, just open the old boxes, set the values with finampSetter, delete the old boxes, and then set the migration flag.  I personally wouldn't bother putting this in its own function and would keep it inside _setupHive, but that's a style choice.

likhithpraveenk committed:
migration: theme mode and locale

likhithpraveenk committed:
add color picker

likhithpraveenk commented:
found this awesome [color picker](https://rydmike.com/flexcolorpicker/) 


<img width="200" height="300" alt="Screenshot_1760043939" src="https://github.com/user-attachments/assets/af16e9a1-9065-436e-a197-e3e13f844f6d" />

likhithpraveenk committed:
add migration flag for themeMode & locale

likhithpraveenk commented:
Any other changes? @Komodo5197 @Chaphasilor

likhithpraveenk committed:
proper extensions

likhithpraveenk committed:
helpful comments

Komodo5197 committed:
Correct issues with background color when not using background images.  Attempt to improve playback action row contrast.

Komodo5197 commented:
The code looks good to me.  I did notice that collection tiles are incorrectly setting a background color, which was hidden by that color matching the background in our themes.  You should remove the tileColor setting in the ListTile in item_collection_list_tile.dart to fix this. It would be nice if you extended the accent picker open button to cover the hex and not just the color, because I kept clicking that.  I think it would be easier to dial in a theme if there was a preview in the color picker, because what a theme will look like is not always obvious.  I think the cleanest way to do this would be to retheme the color picker sheet itself with the preview color, but you could also add a preview widget.

Also, while looking over this I noticed a few issues with how our existing theming code worked when player background images were disabled, and because you had changes to related areas I've taken the liberty of pushing onto your branch.  I hope you don't mind.

Komodo5197 committed:
Store cached themes for all tracks, not just downloaded ones.

likhithpraveenk committed:
apply preview theme to bottom sheet

likhithpraveenk committed:
format

Komodo5197 committed:
Fix withColorScheme not applying text theme changes.

Komodo5197 commented:
While looking over the changes I noticed that I forgot to apply the text color changes during withColorScheme, so I've fixed that.  I noticed you were passing a WidgetRef into the picker modal, which I'd consider bad practice because that ref can't rebuild the picker, so I moved all that code out into a ConsumerStatefulWidget.  Anyway, everything looks ready to merge from my side.  @Chaphasilor Any thoughts?

Chaphasilor committed:
fix log colors, improve wording

Chaphasilor commented:
Changes look good to me, thanks to both of you. I've noticed that the accent color-based color scheme was also used on the logs view, which didn't work well, so I've hard-coded the signal colors there.  
I'll merge this once the checks have passed :)
